lib/postgrex/replication.ex

```diff
@@ -175,53 +183,83 @@ defmodule Postgrex.Replication do
 
   @doc """
-  Receives `binary` replication messages.
+  Invoked after connecting.
 
-  The format of the messages are described [under the START_REPLICATION
-  section in PostgreSQL docs](https://www.postgresql.org/docs/14/protocol-replication.html).
-
-  Some messages require explicit acknowledgement, which can be done
-  by returning a list of binaries according to the replication protocol.
+  This may be invoked multiple times if `:auto_reconnect` is set to true.
   """
-  @callback handle_data(binary, state) :: {:noreply, [copy], state}
+  @callback handle_connect(state) ::
+              {:noreply, ack, state}
+              | {:query, query, state}
+              | {:stream, query, stream_opts, state}
 
   @doc """
-  Receives copy messages.
+  Callback for `:stream` outputs.
 
-  The format of the message is `Postgrex.Result.t()` with the row values
-  formatted as `text`. A `{:copy,done, table_name}` message will be sent
-  to signal copying has completed.
+  If any callback returns `{:stream, iodata, opts, state}`, then this
+  callback will be eventually called with the result of the query.
+  It receives `binary` streaming messages.
+
+  This can be useful for replication and copy commands. For replication,
+  the format of the messages are described [under the START_REPLICATION
+  section in PostgreSQL docs](https://www.postgresql.org/docs/14/protocol-replication.html).
+  Replication messages may require explicit acknowledgement, which can
+  be done by returning a list of binaries according to the replication
+  protocol.
   """
-  @callback handle_copy(Postgrex.Result.t() | {:copy_done, String.t()}, state) ::
-              {:noreply, [copy], state}
+  @callback handle_data(binary, state) ::
+              {:noreply, ack, state}
+              | {:query, query, state}
+              | {:stream, query, stream_opts, state}
 
   @doc """
   Callback for `Kernel.send/2`.
   """
-  @callback handle_info(term, state) :: {:noreply, [copy], state}
+  @callback handle_info(term, state) ::
+              {:noreply, ack, state}
+              | {:query, query, state}
+              | {:stream, query, stream_opts, state}
 
   @doc """
   Callback for `call/3`.
 
   Replies must be sent with `reply/2`.
   """
-  @callback handle_call(term, GenServer.from(), state) :: {:noreply, [copy], state}
-
-  @optional_callbacks handle_info: 2, handle_call: 3, handle_copy: 2
+  @callback handle_call(term, GenServer.from(), state) ::
+              {:noreply, ack, state}
+              | {:query, query, state}
+              | {:stream, query, stream_opts, state}
 
   @doc """
-  Replies to the given client.
+  Callback for `:query` outputs.
 
-  Wrapper for `GenServer.reply/2`.
+  If any callback returns `{:query, iodata, state}`,
+  then this callback will be immediatelly called with
+  the result of the query.
+  """
+  @callback handle_result(%Postgrex.Result{} | %Postgrex.Error{}, state) ::
+              {:noreply, ack, state}
```

Comment:
```suggestion
              {:noreply, state}
```

Would you want to change it to this, and then add another clause in the handler:

```elixir
{:noreply, mod_state} ->
  {:noreply, %{s | state: {mod, mod_state}}}
```

I don't believe any regular queries require ack. I'm pretty sure it's only when you need to let the server know your replication status.

Reply:
Done!